Jupyter for Analysts - Notebook 2

# Notebook 2 already?! Let's do this!

<div class='row'>
   <div class='col-md-5'> 
    <br>       
    <br>
    <br>
    <br>
    <br>
    <br>Now that you have officially ran your first notebook (claps enthusiatically), let's run a notebook that performs some type of data analysis. You know... something your boss really wants you to do.
    <br>
       <br>If you are wild, use the <strong> Shift + Enter </strong> shortcut to run the cell you are in and move to the next cell at the same time!
    </div>
    <div class='col-md-7'>
        <img src="J4A_Pics/that_great.jpg" />
    </div>
</div>
<hr>

<h2>Dave: The Analyst</h2>

One day, an analyst named Dave was looking through his <i>That's What I Call Music</i> spotify playlist. He realized that there are A LOT of different artists over these 1228 songs! A curious analyst, Dave wondered which artist appeared the most throughout these songs.
<br>
<br> Dave went searching for a <i>'That's What I Call Music'</i> data set online. Low and behold, he found one that was perfect to use with Jupyter notebooks!

<h3> What does Dave do next with his data? </h3>
<br>Dave downloaded his dataset to his Jupyter notebook and displayed it in a table using the code below!
<br>
<br> <strong> Hey, you try!! </strong> 
<br> Run the code cell below. It might take a minute or two to 'import' everything. When the cell is done running, the asterik will turn in to a number on the left hand side AND a table of data will display below.
<br>
<br> <i> **** Don't worry about what any of this code means! Just run the cell wait until you see a 'table' displayed in the output! :)</i>

In [1]:
#This code imports a 'package' in python. Don't worry about what this means!
import ipydeps
ipydeps.pip('pandas')
import pandas as pd

##This code reads in a CSV file from nbgallery's cloud, and then displays the data in a table format.
df = pd.read_csv("Datasets/Now_with_Spotify_Final.csv")
data = df[['number', 'volume_number', 'artist', 'title', 'duration']].copy()
data

ModuleNotFoundError: No module named 'ipydeps'

Wowz-ers! Pretty impressive dataset, right? Haven't seen some of these songs in awhile!
<br>
<br> See those dots in the middle of the table? That means there is a lot more data in there, but Jupyter is only showing you the first couple of rows and the last couple of rows. So, this can be a bit overwhelming to have all of that data! I think I will only really need to see the first 5 rows at a time to understand what kind of data I have.
<br>
<br> Run the next code block to see just the first 5 rows of data from the data set.

In [2]:
data.head()

number  volume_number                   artist           title  duration
0       1             13        Justin Timberlake  Rock Your Body       266
1       2             13           Jennifer Lopez        I'm Glad       221
2       3             13                      B2K      Girlfriend       203
3       4             13                    Jay-Z  Excuse Me Miss       229
4       5             13  Ginuwine featuring Baby       Hell Yeah       218

Much better to read and get a gist of the information. Sometimes notebooks will only share a couple of rows of data like this upfront. This doesn't always mean there are only 5 rows, though! 

<br> Ok, so now let's move on to answer <strong>'which artist appears the most over these 61 volumes?'</strong>
<br>
<br> Run the code below to find out the which artist appeared the most on these 61 'That's what I call music' cd's. Read the output yourself and take a guess before looking at the next cell (which has the answer in it!)

In [3]:
#This code take the 'data' table, specifies it wants to look at the artist column, 
#then counts the unique number of times an artist appears.
data.artist.value_counts()

Britney Spears                              18
Katy Perry                                  14
Taylor Swift                                14
Rihanna                                     12
Nickelback                                  12
                                            ..
Nelly featuring Paul Wall and Ali & Gipp     1
T-Pain featuring Yung Joc                    1
The Ting Tings                               1
Celine Dion                                  1
Clairity                                     1
Name: artist, Length: 757, dtype: int64

Once Dave ran the code, he said "Wow! I listen to all these cd's all the time! I didn't know I listened to <strong>Britney Spears</strong> that much! I wonder what else I can do with a Jupyter notebook..."
<hr>

Run the code below to see some basic statistics based off of the <b> duration </b> column from our data.

In [4]:
data.duration.describe()

count    1229.000000
mean      221.580960
std        27.120722
min       125.000000
25%       203.000000
50%       221.000000
75%       237.000000
max       353.000000
Name: duration, dtype: float64

Hmm... looking at the output... it looks like the average (mean) duration is about 222 seconds! The longest song (max) in the series is 353 seconds, and the shortest (min) is 125 seconds. *"I wonder what the shortest and longest songs are..."*
<hr>

# Exercise 2: Interacting with a Jupyter Notebook 'code'

Now we know the values for the shortest and longest song... let's use them to find the longest song and the shortest song!
<br>
<br> This time when you run the code, the <b> output </b> is going to ask <b> YOU </b> for input! In the text box, enter the number for the max when running the code for the longest song and enter the number for min when you are asked to find the shortest song. For example, the longest song is <b> 353 </b>.
<br>
<br> <b> You got this: </b>Run the code. Enter the number in the pop-up text box where the output would be. Then press 'Enter'.

Run the code below to look at the longest song:

In [5]:
## this code asks the user to input the length of the longest song
## it then takes the number the user inputs and finds the song that matches that value
x = input('Please enter the length of the LONGEST song: ')
data.loc[data['duration'] == float(x)]

Please enter the length of the LONGEST song: 353


number  volume_number             artist                title  duration
384       3             48  Justin Timberlake  Take Back the Night       353

Looks like JT holds the longest song! It can be found on cd 48, song 3. The title is 'Take Back the Night' and it is 5 min and 55 seconds (353/60). 

Ok. Now it's your turn to do the <b> exact same thing as above </b> except you need to <b>change the number value to the shortest song length</b>. Run the code below to look at the shortest song:

In [6]:
## this code asks the user to input the length of the shortest song
## it then takes the number the user inputs and finds the song that matches that value
y = input('Please enter the length of the SHORTEST song: ')
data.loc[data['duration'] == float(y)]

Please enter the length of the SHORTEST song: 125


number  volume_number         artist  \
293      12             47  Anna Kendrick   

                                    title  duration  
293  Cups (Pitch Perfect's When I'm Gone)       125

Anna Kendrick holds the shortest song! It can be found on cd 47, song 12. The title is 'Cups' and it is 2 min and 7 seconds (125/60). 

<hr>
<div class='alert alert-block alert-info' align='center'>
Congratulations!
<br>
<br>You just interacted with Jupyter notebook code yourself! I knew you could do it ;)
<br>
    <br>Head on over to <b>J4A Notebook 3</b> and see some other cool things Jupyter notebook can do!